sdk/lib/opentelemetry/sdk/internal.rb

```diff
@@ -49,17 +49,18 @@ def valid_value?(value)
       end
 
       def valid_attributes?(owner, kind, attrs)
-        attrs.nil? || attrs.all? do |k, v|
```

Comment:
I don't exactly understand it, but this was the top allocating location in my benchmark. By switching to `each` (presumably, a simpler implementation under the hood?), the allocation was removed: 

```diff
  allocated objects by location
  -----------------------------------
-     1941  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/opentelemetry-sdk-1.3.1/lib/opentelemetry/sdk/internal.rb:52
      1568  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/opentelemetry-sdk-1.3.1/lib/opentelemetry/sdk/trace/tracer_provider.rb:143
       803  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/graphql-2.1.6/lib/graphql/execution/lazy.rb:21
       784  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/context.rb:151
       784  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/context.rb:153
       784  /Users/rmosolgo/.rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/gems/opentelemetry-api-1.2.3/lib/opentelemetry/trace.rb:39
```

Looking at the breakdown by class, I'm guessing that Ruby makes an Array in there, somewhere: 

```diff 
  allocated objects by class
  -----------------------------------
-     6230  Hash
+     5453  Hash
-     2395  Array
      1188  Proc
-      871  String
+      875 String
       785  Thread::Mutex
       784  OpenTelemetry::Context
```

(The difference in Hash allocations is probably best explained by the other part of this PR.)

Reply:
Weirdly, this is failing CI. I'll have to think through the logic, but we're keeping some Event attributes that we should drop.